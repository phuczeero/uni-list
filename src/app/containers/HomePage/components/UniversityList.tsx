import React from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components/macro';

import { selectUniversities, selectLoading } from '../selectors';
import { Card } from 'app/components/Card';
import { Badge } from 'app/components/Badge';
import { Spinner } from 'app/components/Spinner';

interface Props {}

export function UniversityList(props: Props) {
  const universities = useSelector(selectUniversities);
  const loading = useSelector(selectLoading);

  if (loading)
    return (
      <SpinnerWrapper>
        <Spinner />
      </SpinnerWrapper>
    );

  if (universities.length === 0) {
    return <NotFound>No results found.</NotFound>;
  }

  return (
    <Wrapper>
      {universities.map((u, index) => (
        <Cell key={`${u.name}_${index}`}>
          <ResCard
            title={u.name}
            content={
              <div>
                <Badge>{u.country}</Badge>
                <h4>Web pages:</h4>
                <ul>
                  {u.web_pages.map(wp => (
                    <li key={wp}>
                      <a href={wp} target="_blank" rel="noreferrer">
                        {wp}
                      </a>
                    </li>
                  ))}
                </ul>
              </div>
            }
          />
        </Cell>
      ))}
    </Wrapper>
  );
}

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: ${p => -p.theme.space[2]}px;
  margin-top: ${p => p.theme.space[3]}px;
`;

const ResCard = styled(Card)`
  height: 100%;
`;

const Cell = styled.div`
  width: 100%;
  padding: ${p => p.theme.space[2]}px;

  @media (min-width: ${p => p.theme.breakpoints[1]}) {
    width: 50%;
  }

  @media (min-width: ${p => p.theme.breakpoints[2]}) {
    width: 33.33%;
  }
`;

const NotFound = styled.div`
  color: ${p => p.theme.colors.muted};
  font-size: ${p => p.theme.fontSizes[5]}px;
  padding-top: ${p => p.theme.space[4]}px;
`;

const SpinnerWrapper = styled.div`
  text-align: center;
  padding: ${p => p.theme.space[4]}px 0;
`;
