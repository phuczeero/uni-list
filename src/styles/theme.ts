const fonts = {
  body:
    '-apple-system, BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji"',
};
const fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 72];
const space = [0, 4, 8, 16, 32, 64, 128, 256, 512];
const radii = {
  default: 6,
  circle: 99999,
};
const colors = {
  text: '#212529',
  muted: '#dee2e6',
  primary: '#007bff',
  danger: '#dc3545',
  blues: {
    light: '#dcedff',
    dark: '#004896',
  },
};
const sizes = {
  navbar: 64,
  spinner: 32,
};
const shadows = {
  card: '0 0 4px rgba(0, 0, 0, .125)',
};
const breakpoints = ['576px', '768px', '992px', '1200px'];

export const theme = {
  fonts,
  fontSizes,
  space,
  radii,
  colors,
  sizes,
  shadows,
  breakpoints,
};

export type ThemeType = typeof theme;
