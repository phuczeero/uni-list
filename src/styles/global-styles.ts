import { createGlobalStyle } from 'styled-components';
import { ThemeType } from './theme';

export const GlobalStyle = createGlobalStyle<{ theme: ThemeType }>`
  * {
    font-family: ${p => p.theme.fonts.body};
    color: ${p => p.theme.colors.text}
  }
`;
