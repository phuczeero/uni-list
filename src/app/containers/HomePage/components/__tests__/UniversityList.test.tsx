import * as React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from 'styles/theme';
import { UniversityList } from '../UniversityList';
import { Provider } from 'react-redux';
import { configureAppStore } from 'store/configureStore';

const renderUniversityList = store =>
  render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <UniversityList />
      </ThemeProvider>
    </Provider>,
  );

describe('<UniversityList />', () => {
  let store: ReturnType<typeof configureAppStore>;
  beforeEach(() => {
    store = configureAppStore();
  });

  it('should render spinner', () => {
    const mockStore = {
      ...store,
      getState: () => ({
        homePage: {
          loading: true,
        },
      }),
    };
    const component = renderUniversityList(mockStore);
    expect(component.container.firstChild).toMatchSnapshot();
  });

  it('should render not found message', () => {
    const mockStore = {
      ...store,
      getState: () => ({
        homePage: {
          loading: false,
          universities: [],
        },
      }),
    };
    const component = renderUniversityList(mockStore);
    expect(component.container.firstChild).toMatchSnapshot();
  });

  it('should render university list', () => {
    const mockStore = {
      ...store,
      getState: () => ({
        homePage: {
          loading: false,
          universities: [{ name: 'test', web_pages: ['test.name'] }],
        },
      }),
    };
    const component = renderUniversityList(mockStore);
    expect(component.container.firstChild).toMatchSnapshot();
  });
});
