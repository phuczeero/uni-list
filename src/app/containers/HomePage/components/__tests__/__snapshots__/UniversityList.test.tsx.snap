// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UniversityList /> should render not found message 1`] = `
.c0 {
  color: #dee2e6;
  font-size: 32px;
  padding-top: 32px;
}

<div
  class="c0"
>
  No results found.
</div>
`;

exports[`<UniversityList /> should render spinner 1`] = `
.c1 {
  -webkit-animation: cqkDSr 2.625s linear infinite;
  animation: cqkDSr 2.625s linear infinite;
  height: 32px;
  width: 32px;
  -webkit-transform-origin: center;
  -ms-transform-origin: center;
  transform-origin: center;
}

.c2 {
  -webkit-animation: caVMmK 1.5s ease-in-out infinite;
  animation: caVMmK 1.5s ease-in-out infinite;
  stroke: #007bff;
  stroke-linecap: round;
}

.c0 {
  text-align: center;
  padding: 32px 0;
}

<div
  class="c0"
>
  <svg
    class="c1"
    viewBox="-24 -24 48 48"
  >
    <circle
      class="c2"
      cx="0"
      cy="0"
      fill="none"
      r="20"
      stroke-width="4"
    />
  </svg>
</div>
`;

exports[`<UniversityList /> should render university list 1`] = `
.c2 {
  padding: 16px;
  box-shadow: 0 0 4px rgba(0,0,0,.125);
}

.c4 {
  font-size: 24px;
  font-weight: bold;
}

.c5 {
  padding: 4px;
  background-color: #007bff;
  color: white;
  font-size: 12px;
  border-radius: 99999px;
  display: inline-block;
  min-width: 64px;
  text-align: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin: -8px;
  margin-top: 16px;
}

.c3 {
  height: 100%;
}

.c1 {
  width: 100%;
  padding: 8px;
}

@media (min-width:768px) {
  .c1 {
    width: 50%;
  }
}

@media (min-width:992px) {
  .c1 {
    width: 33.33%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        test
      </div>
      <div>
        <div>
          <div
            class="c5"
          />
          <h4>
            Web pages:
          </h4>
          <ul>
            <li>
              <a
                href="test.name"
                rel="noreferrer"
                target="_blank"
              >
                test.name
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
